src/adapters/methods/reft.py

```diff
@@ -0,0 +1,184 @@
+from .adapter_layer_base import AdapterLayerBase
+from .modeling import Activation_Function_Class
+
+
+class ReftUnit(nn.Module):
+    def __init__(
+        self,
+        in_dim: int,
+        r_dim: int,
+        orthogonal: bool = False,
+        subtract_projection: bool = True,
+        non_linearity: str = None,
+        dropout: float = 0.0,
+    ):
+        super().__init__()
+        self.orthogonal = orthogonal
+        self.learned_source = nn.Linear(in_dim, r_dim, bias=True)
+
+        projection = nn.Linear(in_dim, r_dim, bias=False)
+        if orthogonal:
+            self.projection = nn.utils.parametrizations.orthogonal(projection)
+        else:
+            self.projection = projection
+
+        self.subtract_projection = subtract_projection
+        self.non_linearity = Activation_Function_Class(non_linearity)
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, x):
+        source_states = self.non_linearity(self.learned_source(x))
+        if self.subtract_projection:
+            projected_states = self.projection(x)
+            source_states = source_states - projected_states
+        adapted_output = x + torch.matmul(source_states, self.projection.weight)
+        adapted_output = self.dropout(adapted_output)
+        return adapted_output
+
+
+class ReftModule(nn.Module):
+    def __init__(self, in_features: int, config: ReftConfig):
+        super().__init__()
+        self.prefix_positions = config.prefix_positions
+        self.suffix_positions = config.suffix_positions
+        self.tied_weights = config.tied_weights
+        n_units = 1 if config.tied_weights else 2
+        self.units = nn.ModuleList(
+            [
+                ReftUnit(
+                    in_features,
+                    config.r,
+                    config.orthogonality,
+                    config.subtract_projection,
+                    config.non_linearity,
+                    config.dropout,
+                )
+                for _ in range(n_units)
+            ]
+        )
+
+    def _gather_adapted_states(self, hidden_states: torch.Tensor):
+        # get last non-padding token id
+        context = ForwardContext.get_context()
+        if hasattr(context, "seqlens"):
+            last_non_padding = context.seqlens - 1
+        else:
+            last_non_padding = torch.tensor([hidden_states.size(1) - 1] * hidden_states.size(0))
+        # extract prefix and suffix of seq len
+        pref_ids = self.prefix_positions
+        suff_ids = last_non_padding - self.suffix_positions + 1
+        prefix = hidden_states[:, :pref_ids, :]
+        suffix = []
+        for i, suff_id in enumerate(suff_ids):
+            suffix.append(hidden_states[i, suff_id : suff_id + self.suffix_positions, :])
+        suffix = torch.stack(suffix, dim=0)
+        if self.tied_weights:
+            adapted_states = [torch.cat([prefix, suffix], dim=1)]
+        else:
+            adapted_states = [prefix, suffix]
```

Comment:
Regarding optimizing the reft implementation:
We often have 0 suffix_positions (at least zero is the default in the configs). Couldn't we skip these calculations in these cases?

In general, I don't see how we could optimize `ReftUnit` anymore. So the only methods to optimize are `_gather_adapted_states` and `_scatter_adapted_states`.